### PR TITLE
Story/tao 1227 guest login

### PR DESCRIPTION
**Pull request by bursonic**
Added guest assessment access posibility

Related PRs: 
- oat-sa/tao-core#513
- oat-sa/extension-tao-mp#12
- oat-sa/generis#94

### CODE REVIEW

File: scripts/update/Updater.php [hunk 1]
```diff
@@ -90,6 +90,21 @@ public function update($initialVersion) {
...
+
+            $ext = \common_ext_ExtensionsManager::singleton()->getExtensionById('tao');
+            $loginFormSettings = $ext->getConfig('loginForm');
+            if( empty($loginFormSettings) ){
+                $loginFormSettings = array();
+            }
+
+            $loginFormSettings['elements']['guestAccessLink'] = \taoDelivery_helper_Delivery::getGuestAccessLoginFormElement();
+            $ext->setConfig('loginForm', $loginFormSettings);
+
```
Comment (antoinerobin) [inline scripts/update/Updater.php]: You should grant the access to taoDelivery/DeliveryServer/guest to AnonymousRole

Comment (bursonic) [inline scripts/update/Updater.php]: It's done in the manifest.php here:
https://github.com/oat-sa/extension-tao-delivery/pull/100/files#diff-ff803b7cf2d150cf1cc1a4195c24d2a5R116

Comment (antoinerobin) [inline scripts/update/Updater.php]: The line you added in the manifest only grant permission on install.
To grant permission on update you should add something like that : https://github.com/oat-sa/extension-tao-mediamanager/blob/fix/import-right/scripts/update/Updater.php#L171-L175

Comment (bursonic) [inline scripts/update/Updater.php]: done

File: models/classes/class.GuestTestUser.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Created by PhpStorm.
+ * User: ksasim
+ * Date: 29.8.15
+ * Time: 14.46
+ */
+use oat\oatbox\user\User;
+
+class taoDelivery_models_classes_GuestTestUser implements User
```
Comment (antoinerobin) [inline models/classes/class.GuestTestUser.php]: should better extends common_user_User than implements User cause common_session_BasicSession need a common_user_User in its constructor

Comment (bursonic) [inline models/classes/class.GuestTestUser.php]: I think that better approach is to make constructor accept interface instead of common_user_User

Comment (bursonic) [inline models/classes/class.GuestTestUser.php]: Extend is ok for 'real' ACL users but not for 'fake' users like LTI and Guest

**Comment by antoinerobin** [comment]
as discussed with Patrick the guest login link should appear only when a deployment config is set so the addGuestLink should not be execute for each instance

**Comment by bursonic** [comment]
So which way is the best? I'm just not sure about it.

I can remove this config from taoDelivery install/update scripts.
Then it can be configured manually in loginForm config like this:

```php
'elements' => array(
        'guestAccessLink' => \taoDelivery_helper_Delivery::getGuestAccessLoginFormElement(),
    )
```

**Comment by antoinerobin** [comment]
you should move the install/update scripts to the client repository, like that it will add the guest button only for them

File: actions/class.DeliveryServer.php [hunk 3]
```diff
@@ -83,7 +83,12 @@ public function index(){
 		
 		$deliveryData = array();
 		if (!is_null($user)) {
-		    $available = taoDelivery_models_classes_AssignmentService::singleton()->getAvailableDeliveries($user);
+			$isGuest = $this->service->isDeliveryGuestUser($user);
+
+		    $available = $isGuest
+				? $this->service->getGuestAccessDeliveries()
+				: taoDelivery_models_classes_AssignmentService::singleton()->getAvailableDeliveries($user);
+
```
Comment (antoinerobin) [inline actions/class.DeliveryServer.php]: I would have leave this method has it was and but the guest access logic in the AssignementService. Like that you can regroup all user related method in this service and leave the DeliveryServerService untouched.

File: models/classes/class.DeliveryServerService.php [hunk 4]
```diff
@@ -61,21 +61,59 @@ public function getResumableDeliveries($user = null)
...
         
         $propMaxExec = current($deliveryProps[TAO_DELIVERY_MAXEXEC_PROP]);
         $propStartExec = current($deliveryProps[TAO_DELIVERY_START_PROP]);
         $propEndExec = current($deliveryProps[TAO_DELIVERY_END_PROP]);
+        $propAccessSettings = current($deliveryProps[TAO_DELIVERY_ACCESS_SETTINGS_PROP]);
 
         $settings[TAO_DELIVERY_MAXEXEC_PROP] = (!(is_object($propMaxExec)) or ($propMaxExec=="")) ? 0 : $propMaxExec->literal;
         $settings[TAO_DELIVERY_START_PROP] = (!(is_object($propStartExec)) or ($propStartExec=="")) ? null : $propStartExec->literal;
         $settings[TAO_DELIVERY_END_PROP] = (!(is_object($propEndExec)) or ($propEndExec=="")) ? null : $propEndExec->literal;
+        $settings[TAO_DELIVERY_ACCESS_SETTINGS_PROP] = (!(is_object($propAccessSettings)) or ($propAccessSettings=="")) ? null : $propAccessSettings->getUri();
```
Comment (antoinerobin) [inline models/classes/class.DeliveryServerService.php]: Even if you have this information you don't use it (for now) so you may remove the retrieval of TAO_DELIVERY_ACCESS_SETTINGS_PROP unless you really need it

**Comment by bursonic** [comment]
Updated.
- Sesssion refactored in generis
- Update/Install moved to taoAct
- Guest user logic moved to assignment service

File: models/classes/class.GuestTestUser.php (hunk 2 above)
Comment (antoinerobin) [inline models/classes/class.GuestTestUser.php]: So we discussed with Joel for this issue and the best solution is to pass interface to the constructor as you said and and an interface "Refreshable" (let the common_user_User implement it) and check in the refresh method of the basic session if the user implements this interface before call the method on it

**Pull request merged**